README.md

```diff
@@ -0,0 +1,396 @@
+This reduces cognitive load when using components -- no need to remember one component's peculiarities.
+It also allows Agent-wide changes, where we make the same formulaic change to each component.
+If a situation arises that contradicts the guidelines, then we can update the guidelines (and change all affected components).
+
+## Prerequisites
+
+You should be familiar with [`fx`](https://pkg.go.dev/go.uber.org/fx), the tool used for dependency injection.
+
+## Component Definition
+
+A component is defined in a dedicated package under `comp/`, with the following defined in `component.go`:
+
+ * Extensive package-level documentation.
+   This should define, as precisely as possible, the behavior of the component, acting as a contract on which users of the component may depend.
+
+   The documentation (both package-level and method-level) should also include:
+
+   * Precise information about which interface methods can be called during the setup phase, and which must only be called after the component is started.
+   * Precise information about data ownership of passed values and returned values (does the method modify its arguments?  can the caller modify a returned slice or map?).
+   * Precise information about goroutines and blocking (does the method block? is a callback invoked in a dedicated goroutine? what happens if a channel is full?)
+
+ * `pkg.Component` -- the type implemented by the component.
+   This can be an empty interface, but is the type by which other components will find this one via `fx`.
+   It should have a formulaic doc string like `// Component is the component type.`, deferring documentation to the package docs.
+   All interface methods should be exported and thoroughly documented.
+
+ * `pkg.Module` -- an `fx.Option` that can be included in an `fx.App` to make this component available.
+   To improve logging, use `fx.Module(comppath, ..)`, where `comppath` is the root-relative package path, e.g., `"comp/util/log"`.
+   It should have a formulaic doc string like `// Module defines the fx options for this component.`
+
+Components should not be nested; that is, no component's Go path should be a prefix of another component's Go path.
+
+### Implementation
+
+The completed `component.go` looks like this:
+
+```go
+// Package foo ... (detailed doc comment for the component)
+package config
+
+// team: some-team-name
+
+// Component is the component type.
+type Component interface {
+	// Foo is ... (detailed doc comment)
+	Foo(key string) string
+}
+
+// Module defines the fx options for this component.
+var Module = fx.Module(
+    "comp/foo", // (package path of the component)
+    fx.Provide(newFoo),
+)
+```
+
+The Component interface is implemented in another file by an unexported type with a sensible name such as `launcher` or `provider`.
+
+```go
+package config
+
+type foo {
+    foos []string
+}
+
+// Foo implements Component#Foo.
+func (f *foo) Foo(key string) string { ... }
+
+type dependencies struct {
```

Comment:
I think I softened that in the paragraph below?  I suspect that `dependencies` is a good idea for most components, but for the simplest components it may be omitted.

Reply:
Agreed, using one struct for dependencies most of the time a good idea (when there are many dependencies) but let the possibility to not use it when there is a single dependencies is probably also a good idea.